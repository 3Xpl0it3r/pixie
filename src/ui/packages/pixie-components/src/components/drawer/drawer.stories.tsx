/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as React from 'react';

import Button from '@material-ui/core/Button';
import { FixedSizeDrawer } from './drawer';

export default {
  title: 'Drawer/Fixed',
  component: FixedSizeDrawer,
};

export const Basic = () => {
  const [open, setOpen] = React.useState(false);
  const toggleOpen = React.useCallback(() => setOpen((opened) => !opened), []);

  const otherContent = (
    <div>
      <Button onClick={toggleOpen} color='primary'>
        {open ? 'Close' : 'Open'}
      </Button>
      <div>Other content. Some text goes here.</div>
    </div>
  );

  return (
    <FixedSizeDrawer
      drawerDirection='left'
      drawerSize='250px'
      open={open}
      otherContent={otherContent}
      overlay={false}
    >
      <div>Drawer contents</div>
    </FixedSizeDrawer>
  );
};
