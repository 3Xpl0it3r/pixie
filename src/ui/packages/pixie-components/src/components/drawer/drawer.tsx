/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as React from 'react';
import Drawer from '@material-ui/core/Drawer';
import { WithStyles, withStyles } from '@material-ui/core/styles';
import { createStyles } from '@material-ui/styles';

export type DrawerDirection = 'top' | 'bottom' | 'left' | 'right';

const styles = () => createStyles({
  root: {
    display: 'flex',
    flex: 1,
    minHeight: 0,
    minWidth: 0,
    position: 'relative',
  },
  otherContent: {
    flex: 1,
    display: 'flex',
    width: '100%',
  },
  drawerContents: {
    display: 'flex',
    width: '100%',
    height: '100%',
  },
  drawerPaper: {
    position: 'absolute',
  },
  dockedVertical: {
    width: 0,
    flex: 0,
  },
  dockedHorizontal: {
    height: 0,
    flex: 0,
  },
});

interface FixedSizeDrawerProps extends WithStyles<typeof styles> {
  children?: React.ReactNode; // The contents of the drawer.
  otherContent?: React.ReactNode; // The content that is not in the drawer.
  drawerDirection: DrawerDirection;
  drawerSize: string; // A fixed size for the drawer.
  open: boolean;
  overlay: boolean;
}

const FixedSizeDrawerImpl = ({
  classes,
  children,
  otherContent,
  drawerDirection,
  drawerSize,
  open,
  overlay,
}: FixedSizeDrawerProps) => {
  const drawerStyle = drawerDirection === 'top' || drawerDirection === 'bottom'
    ? { height: drawerSize }
    : { width: drawerSize };

  let contentStyle = {};
  if (open && !overlay) {
    // When the drawer is open, the other content should shrink by `drawerSize`.
    if (drawerDirection === 'top') {
      contentStyle = { marginTop: drawerSize };
    } else if (drawerDirection === 'bottom') {
      contentStyle = { marginBottom: drawerSize };
    } else if (drawerDirection === 'left') {
      contentStyle = { marginLeft: drawerSize };
    } else {
      contentStyle = { marginRight: drawerSize };
    }
  }

  return (
    <div className={classes.root}>
      <div className={classes.otherContent} style={contentStyle}>
        {otherContent}
      </div>
      <Drawer
        anchor={drawerDirection}
        style={drawerStyle}
        variant='persistent'
        open={open}
        classes={{
          paper: classes.drawerPaper,
          docked:
            drawerDirection === 'top' || drawerDirection === 'bottom'
              ? classes.dockedVertical
              : classes.dockedHorizontal,
        }}
      >
        <div className={classes.drawerContents} style={drawerStyle}>
          {children}
        </div>
      </Drawer>
    </div>
  );
};
FixedSizeDrawerImpl.defaultProps = {
  children: null,
  otherContent: null,
};

export const FixedSizeDrawer = withStyles(styles)(FixedSizeDrawerImpl);
