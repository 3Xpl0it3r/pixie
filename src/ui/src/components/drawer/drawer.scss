.pixie-drawer {
  display: flex;
  flex-direction: column;
}

.pixie-drawer-content {
  flex: 1;
  min-height: 0;

  &.closed > * {
    display: none;
  }
}

.pixie-drawer-footer-row {
  display: flex;
  flex-direction: row;
  background-color: inherit;
  height: 2rem;
}
