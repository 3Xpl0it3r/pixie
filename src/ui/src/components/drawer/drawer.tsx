// @ts-ignore : TS does not like image files.
import * as collapseLeft from 'images/icons/collapse-left.svg';
// @ts-ignore : TS does not like image files.
import * as collapseRight from 'images/icons/collapse-right.svg';
import * as React from 'react';
import {Button} from 'react-bootstrap';

interface DrawerProps {
  defaultOpened?: boolean;
  openedWidth?: string;
  closedWidth?: string;
}

export const Drawer = React.memo<React.PropsWithChildren<DrawerProps>>(
  ({
    children,
    openedWidth = '10rem',
    closedWidth = '2rem',
    defaultOpened = true,
  }) => {
    const [opened, setOpened] = React.useState<boolean>(defaultOpened);
    const toggleOpened = () => setOpened((isOpened) => !isOpened);

    return (
      <div
        className={`pixie-drawer ${opened ? 'opened' : 'closed'}`}
        style={{
          display: 'flex',
          flexDirection: 'column',
          width: opened ? openedWidth : closedWidth,
        }}
      >
        <div
          className='pixie-drawer-content'
          style={{
            flex: 1,
            ...(opened ? {} : { visibility: 'hidden' }),
          }}
        >
          {children}
        </div>
        <div style={{ display: 'flex', flexDirection: 'row' }}>
          <div style={{ flex: 1 }} />
          <Button size='sm' onClick={toggleOpened}>
            <img src={opened ? collapseLeft : collapseRight} />
          </Button>
        </div>
      </div>
    );
  });
