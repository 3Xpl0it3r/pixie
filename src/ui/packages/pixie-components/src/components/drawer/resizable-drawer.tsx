import * as React from 'react';
import { DraggableCore } from 'react-draggable';
import {
  createStyles, Theme, WithStyles, withStyles,
} from '@material-ui/core/styles';

import { FixedSizeDrawer, DrawerDirection } from './drawer';

// The amount of time elasped since a user has last resized the drawer.
const RESIZE_WAIT_INTERVAL_MS = 100;

const styles = ({ spacing }: Theme) => createStyles({
  draggableContent: {
    flex: 1,
    minHeight: 0,
    minWidth: 0,
  },
  verticalDragHandle: {
    cursor: 'row-resize',
    width: '100%',
    height: spacing(1),
    zIndex: 1500,
    position: 'absolute',
  },
  horizontalDragHandle: {
    cursor: 'col-resize',
    height: '100%',
    width: spacing(1),
    zIndex: 1500,
    position: 'absolute',
  },
});

interface ResizableDrawerProps extends WithStyles<typeof styles> {
  children?: React.ReactNode; // The contents of the drawer.
  otherContent?: React.ReactNode; // The content that is not in the drawer.
  drawerDirection: DrawerDirection;
  open: boolean;
  initialSize: number;
  overlay: boolean;
}

const ResizableDrawerImpl = ({
  classes, children, otherContent, drawerDirection, open, overlay, initialSize,
}: ResizableDrawerProps) => {
  const [drawerSize, setDrawerSize] = React.useState(initialSize);
  // State responsible for tracking whether the user is actively resizing. This is used for debouncing.
  const [timer, setTimer] = React.useState(null);

  const resize = React.useCallback(({ deltaY, deltaX }) => {
    clearTimeout(timer);
    const newTimer = setTimeout(() => {
      window.dispatchEvent(new Event('resize'));
    }, RESIZE_WAIT_INTERVAL_MS);
    setTimer(newTimer);

    switch (drawerDirection) {
      case 'top':
        setDrawerSize(drawerSize + deltaY);
        break;
      case 'bottom':
        setDrawerSize(drawerSize - deltaY);
        break;
      case 'left':
        setDrawerSize(drawerSize + deltaX);
        break;
      case 'right':
        setDrawerSize(drawerSize - deltaX);
        break;
      default:
        break;
    }
  }, [drawerDirection, drawerSize, timer]);

  let dragHandleStyle = {};
  // If the drawer is in overlay mode, we need additional styling to place the drag
  // handle in the correct place, since we can't rely on the drawer being in the
  // DOM anymore.
  if (overlay) {
    switch (drawerDirection) {
      case 'top':
        dragHandleStyle = { left: 0, marginTop: drawerSize };
        break;
      case 'bottom':
        dragHandleStyle = { left: 0, bottom: 0, marginBottom: drawerSize };
        break;
      case 'left':
        dragHandleStyle = { top: 0, marginLeft: drawerSize };
        break;
      case 'right':
        dragHandleStyle = { top: 0, right: 0, marginRight: drawerSize };
        break;
      default:
        break;
    }
  }

  const dragHandle = (
    <DraggableCore
      onDrag={(_, { deltaX, deltaY }) => {
        resize({
          deltaY,
          deltaX,
        });
      }}
    >
      <div
        style={dragHandleStyle}
        className={drawerDirection === 'top' || drawerDirection === 'bottom'
          ? classes.verticalDragHandle : classes.horizontalDragHandle}
      />
    </DraggableCore>
  );

  let dragFlex = {};
  if (drawerDirection === 'left' || drawerDirection === 'right') {
    dragFlex = { display: 'flex', flexDirection: 'column' };
  }

  const draggableContent = (
    <div className={classes.draggableContent} style={dragFlex}>
      {
        (drawerDirection === 'top' || drawerDirection === 'left') && open
        && dragHandle
      }
      {otherContent}
      {
        (drawerDirection === 'bottom' || drawerDirection === 'right') && open
        && dragHandle
      }
    </div>
  );

  return (
    <>
      <FixedSizeDrawer
        otherContent={draggableContent}
        drawerDirection={drawerDirection}
        open={open}
        overlay={overlay}
        drawerSize={`${drawerSize}px`}
      >
        {children}
      </FixedSizeDrawer>
    </>
  );
};

export const ResizableDrawer = withStyles(styles)(ResizableDrawerImpl);
