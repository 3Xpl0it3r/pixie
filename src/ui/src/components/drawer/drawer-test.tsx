import * as React from 'react';
import { render } from 'enzyme';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core';
import { FixedSizeDrawer } from './drawer';

// TODO(michelle): Currently doesn't work with the clsx import. Enable when we've solved that problem.
xdescribe('<FixedSizeDrawer/>', () => {
  it('renders correctly when closed', () => {
    const otherContent = (
      <div>
        Other content. Some text goes here.
      </div>
    );

    const wrapper = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <FixedSizeDrawer
          drawerDirection='left'
          drawerSize='50px'
          open={false}
          otherContent={otherContent}
        >
          <div>
            Drawer contents
          </div>
        </FixedSizeDrawer>
      </MuiThemeProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
