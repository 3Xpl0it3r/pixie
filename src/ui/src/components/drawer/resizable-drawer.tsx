import * as React from 'react';
import { DraggableCore } from 'react-draggable';
import {
  createStyles, Theme, WithStyles, withStyles,
} from '@material-ui/core/styles';

import { FixedSizeDrawer, DrawerDirection } from './drawer';

const styles = ({ spacing }: Theme) => createStyles({
  draggableContent: {
    flex: 1,
    minHeight: 0,
    minWidth: 0,
  },
  verticalDragHandle: {
    cursor: 'row-resize',
    width: '100%',
    height: spacing(1),
    zIndex: 1500,
    position: 'absolute',
  },
  horizontalDragHandle: {
    cursor: 'col-resize',
    height: '100%',
    width: spacing(1),
    zIndex: 1500,
    position: 'absolute',
  },
});

interface ResizableDrawerProps extends WithStyles<typeof styles> {
  children?: React.ReactNode; // The contents of the drawer.
  otherContent?: React.ReactNode; // The content that is not in the drawer.
  drawerDirection: DrawerDirection;
  open: boolean;
  initialSize: number;
  overlay: boolean;
}

const ResizableDrawer = ({
  classes, children, otherContent, drawerDirection, open, overlay, initialSize,
}: ResizableDrawerProps) => {
  const [drawerSize, setDrawerSize] = React.useState(initialSize);

  const resize = React.useCallback(({ deltaY, deltaX }) => {
    switch (drawerDirection) {
      case 'top':
        setDrawerSize(drawerSize + deltaY);
        break;
      case 'bottom':
        setDrawerSize(drawerSize - deltaY);
        break;
      case 'left':
        setDrawerSize(drawerSize + deltaX);
        break;
      case 'right':
        setDrawerSize(drawerSize - deltaX);
        break;
      default:
        break;
    }
  }, [drawerDirection, drawerSize]);

  let dragHandleStyle = {};
  // If the drawer is in overlay mode, we need additional styling to place the drag
  // handle in the correct place, since we can't rely on the drawer being in the
  // DOM anymore.
  if (overlay) {
    switch (drawerDirection) {
      case 'top':
        dragHandleStyle = { left: 0, marginTop: drawerSize };
        break;
      case 'bottom':
        dragHandleStyle = { left: 0, bottom: 0, marginBottom: drawerSize };
        break;
      case 'left':
        dragHandleStyle = { top: 0, marginLeft: drawerSize };
        break;
      case 'right':
        dragHandleStyle = { top: 0, right: 0, marginRight: drawerSize };
        break;
      default:
        break;
    }
  }

  const dragHandle = (
    <DraggableCore
      onDrag={(event, { deltaX, deltaY }) => {
        resize({
          deltaY,
          deltaX,
        });
      }}
    >
      <div
        style={dragHandleStyle}
        className={drawerDirection === 'top' || drawerDirection === 'bottom'
          ? classes.verticalDragHandle : classes.horizontalDragHandle}
      />
    </DraggableCore>
  );

  const draggableContent = (
    <div className={classes.draggableContent}>
      {
        (drawerDirection === 'top' || drawerDirection === 'left') && open
        && dragHandle
      }
      {otherContent}
      {
        (drawerDirection === 'bottom' || drawerDirection === 'right') && open
        && dragHandle
      }
    </div>
  );

  return (
    <>
      <FixedSizeDrawer
        otherContent={draggableContent}
        drawerDirection={drawerDirection}
        open={open}
        overlay={overlay}
        drawerSize={`${drawerSize}px`}
      >
        {children}
      </FixedSizeDrawer>
    </>
  );
};

export default withStyles(styles)(ResizableDrawer);
