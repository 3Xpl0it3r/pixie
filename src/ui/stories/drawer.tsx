import * as React from 'react';

import { storiesOf } from '@storybook/react';
import Button from '@material-ui/core/Button';
import { FixedSizeDrawer } from '../src/components/drawer/drawer';
import ResizableDrawer from '../src/components/drawer/resizable-drawer';

storiesOf('Drawer', module)
  .add('FixedSizeDrawer', () => {
    const [open, setOpen] = React.useState(false);
    const toggleOpen = React.useCallback(() => setOpen((opened) => !opened), []);

    const otherContent = (
      <div>
        <Button onClick={toggleOpen} color='primary'>
          { open ? 'Close' : 'Open'}
        </Button>
        <div>
          Other content. Some text goes here.
        </div>
      </div>
    );

    return (
      <div>
        <FixedSizeDrawer
          drawerDirection='left'
          drawerSize='50px'
          open={open}
          otherContent={otherContent}
        >
          <div>
            Drawer contents
          </div>
        </FixedSizeDrawer>
      </div>
    );
  }, {
    info: { inline: true },
    notes: 'This is the basic fixed-size drawer, opening from the left',
  }).add('ResizableDrawer', () => {
    const [open, setOpen] = React.useState(false);
    const toggleOpen = React.useCallback(() => setOpen((opened) => !opened), []);

    const otherContent = (
      <div>
        <Button onClick={toggleOpen} color='primary'>
          { open ? 'Close' : 'Open'}
        </Button>
        <div>
          Other content. Some text goes here.
        </div>
      </div>
    );

    return (
      <div>
        <ResizableDrawer
          drawerDirection='left'
          open={open}
          otherContent={otherContent}
          initialSize={200}
        >
          <div>
            Drawer contents
          </div>
        </ResizableDrawer>
      </div>
    );
  }, {
    info: { inline: true },
    notes: 'This is the basic resizable drawer, opening from the left',
  });
