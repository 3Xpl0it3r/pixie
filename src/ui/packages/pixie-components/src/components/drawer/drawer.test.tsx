import * as React from 'react';
import { render } from 'enzyme';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core';
import { FixedSizeDrawer } from './drawer';

describe('<FixedSizeDrawer/>', () => {
  it('renders correctly when closed', () => {
    const otherContent = <div>Other content. Some text goes here.</div>;

    const wrapper = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <FixedSizeDrawer
          drawerDirection='left'
          drawerSize='50px'
          open={false}
          otherContent={otherContent}
          overlay={false}
        >
          <div>Drawer contents</div>
        </FixedSizeDrawer>
      </MuiThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
