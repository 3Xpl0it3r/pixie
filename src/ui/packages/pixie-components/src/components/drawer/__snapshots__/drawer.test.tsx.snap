// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FixedSizeDrawer/> renders correctly when closed 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "FixedSizeDrawerImpl-root-1",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "FixedSizeDrawerImpl-otherContent-2",
        },
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "Other content. Some text goes here.",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "MuiDrawer-root MuiDrawer-docked FixedSizeDrawerImpl-dockedHorizontal-6",
            "style": "width:50px",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "MuiPaper-root MuiDrawer-paper FixedSizeDrawerImpl-drawerPaper-4 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0",
                "style": "visibility:hidden",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "FixedSizeDrawerImpl-drawerContents-3",
                    "style": "width:50px",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {},
                      "children": Array [
                        Object {
                          "data": "Drawer contents",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {},
                      "x-attribsPrefix": Object {},
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "style": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "style": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "MuiDrawer-root MuiDrawer-docked FixedSizeDrawerImpl-dockedHorizontal-6",
          "style": "width:50px",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiPaper-root MuiDrawer-paper FixedSizeDrawerImpl-drawerPaper-4 MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0",
              "style": "visibility:hidden",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "FixedSizeDrawerImpl-drawerContents-3",
                  "style": "width:50px",
                },
                "children": Array [
                  Object {
                    "attribs": Object {},
                    "children": Array [
                      Object {
                        "data": "Drawer contents",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {},
                    "x-attribsPrefix": Object {},
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "style": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "style": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "FixedSizeDrawerImpl-otherContent-2",
          },
          "children": Array [
            Object {
              "attribs": Object {},
              "children": Array [
                Object {
                  "data": "Other content. Some text goes here.",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
